compiler/noirc_evaluator/src/ssa/opt/flatten_cfg/value_merger.rs

```diff
@@ -175,12 +175,18 @@ impl<'a> ValueMerger<'a> {
             _ => panic!("Expected slice type"),
         };
 
-        let then_len = *self.slice_sizes.get(&then_value_id).unwrap_or_else(|| {
-            panic!("ICE: Merging values during flattening encountered slice {then_value_id} without a preset size");
+        let then_len = self.slice_sizes.get(&then_value_id).copied().unwrap_or_else(|| {
+            let (slice, typ) = self.dfg.get_array_constant(then_value_id).unwrap_or_else(|| {
+                panic!("ICE: Merging values during flattening encountered slice {then_value_id} without a preset size");
+            });
+            slice.len() / typ.element_types().len()
         });
 
-        let else_len = *self.slice_sizes.get(&else_value_id).unwrap_or_else(|| {
-            panic!("ICE: Merging values during flattening encountered slice {else_value_id} without a preset size");
+        let else_len = self.slice_sizes.get(&else_value_id).copied().unwrap_or_else(|| {
+            let (slice, typ) = self.dfg.get_array_constant(else_value_id).unwrap_or_else(|| {
```

Comment:
Why did you have to do this in the case there is no slice size set? The goal with `slice_sizes` was that everything is already precomputed by value merging

Reply:
This PR reimplements slice size tracking so it is a bit different. This check is added so that I don't have to check every instruction for array/slice literals to record their size. I can just track instructions that change the size instead with this check catching any array literals that haven't been caught.

Reply:
Ok makes sense. It was also was more relevant to track literals outside of the value merger when we had nested slices, but not an issue anymore.